proposals/dao/clawback.ts

```diff
@@ -0,0 +1,183 @@
+import hre, { ethers, artifacts } from 'hardhat';
+import { expect } from 'chai';
+import {
+  DeployUpgradeFunc,
+  NamedAddresses,
+  SetupUpgradeFunc,
+  TeardownUpgradeFunc,
+  ValidateUpgradeFunc
+} from '@custom-types/types';
+import { LinearTimelockedDelegator, LinearTokenTimelock, QuadraticTimelockedDelegator } from '@custom-types/contracts';
+import { BigNumber } from 'ethers';
+import { getImpersonatedSigner } from '@test/helpers';
+import { forceEth } from '@test/integration/setup/utils';
+
+const toBN = ethers.BigNumber.from;
+/*
+Migrate the Rari Infra timelocks and return the unvested Jai and David TRIBE tokens
+
+Rari Infra timelocks
+- There is a Fei and a TRIBE timelock
+- Both active from Dec 2021
+- Duration set was 63072000 seconds (730 days, 2 years)
+- They release 2M FEI and 2M TRIBE per year
+- Linear timelocks
+
+Jai and David Vesting contracts
+- 3 vesting contracts with the same terms
+- Active from Dec 2021
+- Quadratic back weighted 5 year vesting
+- There is 17.1M TRIBE total across the 3 vesting contracts
+- The beneficiary of each vesting contract is Gnosis Safe multisig. Owners are unknown
+
+*/
+
+const fipNumber = 'FIP-113: End Departed Rari Founders Vesting of TRIBE';
+
+// TODO: Update these based on the final balances
+const OLD_RARI_TIMELOCK_FEI_AMOUNT = '3254306506849315068493151';
+const OLD_RARI_TIMELOCK_TRIBE_AMOUNT = '3254296867072552004058854';
+
+// Sanity check, minimum amount DAO should receive when clawing back
+const MIN_CLAWED_TRIBE = ethers.constants.WeiPerEther.mul(15_000_000);
+
+// Sanity check, minimum amount clawback vesting contract beneficiary should receive
+const MIN_BENEFICIARY_CLAWED_TRIBE = ethers.constants.WeiPerEther.mul(100_000);
+
+let initialDAOTribeBalance: BigNumber;
+let initialClawbackBeneficiaryBalance: BigNumber;
+let rariFeiTimelockRemainingDuration: BigNumber;
+let rariTribeTimelockRemainingDuration: BigNumber;
+
+// Do any deployments
+// This should exclusively include new contract deployments
+const deploy: DeployUpgradeFunc = async (deployAddress: string, addresses: NamedAddresses, logging: boolean) => {
+  const rariInfraFeiTimelock = await ethers.getContractAt('LinearTokenTimelock', addresses.rariInfraFeiTimelock);
+  const rariInfraTribeTimelock = await ethers.getContractAt(
+    'LinearTimelockedDelegator',
+    addresses.rariInfraTribeTimelock
+  );
+
+  rariFeiTimelockRemainingDuration = await rariInfraFeiTimelock.remainingTime();
+  rariTribeTimelockRemainingDuration = await rariInfraTribeTimelock.remainingTime();
+
+  // 1. Deploy new Rari infra vesting contract
+  const LinearTimelockedDelegatorFactory = await ethers.getContractFactory('LinearTimelockedDelegator'); // change timelock type
+  const newRariInfraFeiTimelock = await LinearTimelockedDelegatorFactory.deploy(
```

Comment:
the FEI timelock should be a `LinearTokenTimelock` there is no `delegate(address)` call on the FEI token contract

Reply:
Ah good point, oversight